sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala

```diff
@@ -116,47 +116,53 @@ object OrcUtils extends Logging {
   }
 
   /**
-   * Returns the requested column ids from the given ORC file. Column id can be -1, which means the
-   * requested column doesn't exist in the ORC file. Returns None if the given ORC file is empty.
+   * @return Returns the requested column ids from the given ORC file and Boolean flag to use actual
+   * schema or result schema. Column id can be -1, which means the requested column doesn't
+   * exist in the ORC file. Returns None if the given ORC file is empty.
    */
   def requestedColumnIds(
       isCaseSensitive: Boolean,
       dataSchema: StructType,
       requiredSchema: StructType,
       reader: Reader,
-      conf: Configuration): Option[Array[Int]] = {
+      conf: Configuration): (Option[Array[Int]], Boolean) = {
+    var sendActualSchema = false
     val orcFieldNames = reader.getSchema.getFieldNames.asScala
```

Comment:
let's say we are running the test `CREATE TABLE test_hive_orc_impl ...`, what's the value of `dataSchema`, `requiredSchema` and `orcFieldNames` when we reach here?

Reply:
For this example query
```
val u = """select date_dim.d_year from date_dim limit 5"""

spark.sql(u).collect
```

value of the orcFieldNames
```
val of orcFieldNames = {Wrappers$JListWrapper@13621} "Wrappers$JListWrapper" size = 28
 0 = "_col0"
 1 = "_col1"
 2 = "_col2"
 3 = "_col3"
 4 = "_col4"
 5 = "_col5"
 6 = "_col6"
 7 = "_col7"
 8 = "_col8"
 9 = "_col9"
 10 = "_col10"
```

value of dataSchema
```
value of dataSchema = {StructType@13616} "StructType" size = 28
 0 = {StructField@16487} "StructField(d_date_sk,IntegerType,true)"
 1 = {StructField@16488} "StructField(d_date_id,StringType,true)"
 2 = {StructField@16489} "StructField(d_date,TimestampType,true)"
 3 = {StructField@16490} "StructField(d_month_seq,IntegerType,true)"
 4 = {StructField@16491} "StructField(d_week_seq,IntegerType,true)"
 5 = {StructField@16492} "StructField(d_quarter_seq,IntegerType,true)"
 6 = {StructField@16493} "StructField(d_year,IntegerType,true)"
 7 = {StructField@16494} "StructField(d_dow,IntegerType,true)"
 8 = {StructField@16495} "StructField(d_moy,IntegerType,true)"
 9 = {StructField@16496} "StructField(d_dom,IntegerType,true)"
 10 = {StructField@16497} "StructField(d_qoy,IntegerType,true)"
```

value of requiredSchema

```
result = {StructType@16990} "StructType" size = 1
 0 = {StructField@17128} "StructField(d_year,IntegerType,true)"
```

Reply:
Please correct me if I'm wrong:
1. the physical orc file schema is `_col0`, ...
2. the table schema in metastore is `d_date_sk`, ...
3. the query only requires only `d_year`

I don't know why the query fails. The `requestedColumnIds` will be `[6]` and the orc reader will read the `_col6` column. Everything should be fine.

Reply:
In this code , this is the place where we are getting exception

https://github.com/apache/spark/blob/d6a68e0b67ff7de58073c176dd097070e88ac831/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java#L183

So the code is creating a wrap VectorizedRowBatchWrap using the result schema which is 
```
result = {StructType@16990} "StructType" size = 1
 0 = {StructField@17128} "StructField(d_year,IntegerType,true)"
```

https://github.com/apache/spark/blob/d6a68e0b67ff7de58073c176dd097070e88ac831/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java#L149 

Now wrap is created with size 1 

and in this code 
` code orcVectorWrappers[i] = new OrcColumnVector(dt, wrap.batch().cols[colId]);`

value of colId will be 6 (which is the requestedColumnIds will be [6])

Now the size of wrap is 1 and code is requesting the fetch wrap.batch().cols[6] so here its getting ArrayIndexOutOfBoundsException

Reply:
Does the failure go away if the physical orc file schema matches the table schema in metastore?

Reply:
yes, if that would be the case then it will be same as how the data created from the spark Application using orc. And it follows the code flow as its working today for spark orc data source tables.

So this only failing for the orc data created by hive, So if I create the data from this using the spark orc datasource , This error is not coming

```
val u = """select * from date_dim limit 5"""

scala> spark.sql(u).write.format("orc").save("/Users/tpcdsdata/testFS/testorc")

val table = """CREATE TABLE `date_dim345` (
     |   `d_date_sk` INT,
     |   `d_date_id` STRING,
     |   `d_date` TIMESTAMP,
     |   `d_month_seq` INT,
     |   `d_week_seq` INT,
     |   `d_quarter_seq` INT,
     |   `d_year` INT,
     |   `d_dow` INT,
     |   `d_moy` INT,
     |   `d_dom` INT,
     |   `d_qoy` INT,
     |   `d_fy_year` INT,
     |   `d_fy_quarter_seq` INT,
     |   `d_fy_week_seq` INT,
     |   `d_day_name` STRING,
     |   `d_quarter_name` STRING,
     |   `d_holiday` STRING,
     |   `d_weekend` STRING,
     |   `d_following_holiday` STRING,
     |   `d_first_dom` INT,
     |   `d_last_dom` INT,
     |   `d_same_day_ly` INT,
     |   `d_same_day_lq` INT,
     |   `d_current_day` STRING,
     |   `d_current_week` STRING,
     |   `d_current_month` STRING,
     |   `d_current_quarter` STRING,
     |   `d_current_year` STRING)
     | USING orc
     | LOCATION '/Users/tpcdsdata/testFS/testorc/'"""

spark.sql(table).collect
val u = """select date_dim.d_date_id from date_dim345 limit 5"""
```
Now this is having the correct value of  physical orc file not _col1, _col2 etc
```
orcFieldNames = {Wrappers$JListWrapper@19940} "Wrappers$JListWrapper" size = 28
 1 = "d_date_id"
 9 = "d_dom"
 7 = "d_dow"
 11 = "d_fy_year"
 5 = "d_quarter_seq"
 0 = "d_date_sk"
 8 = "d_moy"
 10 = "d_qoy"
 4 = "d_week_seq"
 6 = "d_year"
 3 = "d_month_seq"
 2 = "d_date"
```

Reply:
following your previous explanation in https://github.com/apache/spark/pull/29045#discussion_r454234413

1. the code is creating a wrap VectorizedRowBatchWrap using the result schema which is `<d_year, int>`
2. `orcVectorWrappers[i] = new OrcColumnVector(dt, wrap.batch().cols[colId]);` accesses the 6th column of the orc batch.

Why it doesn't fail when the physical orc file schema matches the table schema in metastore?

Reply:
In this case code follows this path

https://github.com/apache/spark/blob/d6a68e0b67ff7de58073c176dd097070e88ac831/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala#L158

Now requiredSchema (result schema) is used for getting the column id instead of entire dataschema . So in this scenario required requiredSchema(requiredSchema) is used to create the VectorizedRowBatchWrap
wrap = new VectorizedRowBatchWrap(orcSchema.createRowBatch(capacity)); 


Where as in failure case column id got from the dataschema and VectorizedRowBatchWrap is created using the requiredSchema (result schema).